_posts/2021-11-23-how-to-dry-your-ref-arch.adoc

```diff
@@ -0,0 +1,588 @@
+https://github.com/gruntwork-io/terraform-aws-service-catalog/tree/master/modules/services/k8s-service[k8s-service]
+Service Catalog Terraform module takes in the container image in the format:
+
+[source,hcl]
+----
+type = object({
+  # Repository of the docker image (e.g. gruntwork/frontend-service)
+  repository = string
+  # The tag of the docker image to deploy.
+  tag = string
+  # The image pull policy. Can be one of IfNotPresent, Always, or Never.
+  pull_policy = string
+})
+----
+
+It is typical for only the `tag` attribute to be different across environments. In the guide, you would have to repeat
+this configuration across the environments, but with deep merge, you can promote the `repository` and `pull_policy`
+attributes to the common component configuration:
+
+.common component configuration
+[source,hcl]
+----
+inputs = {
+  container_image = {
+    repository  = "gruntwork/aws-sample-app"
+    pull_policy = "IfNotPresent"
+  }
+}
+----
+
+.dev configuration
+[source,hcl]
+----
+include "parent" {
+  path           = "/path/to/parent/configuration"
+  merge_strategy = "deep"
+}
+
+inputs = {
+  container_image = {
+    tag = "v0.0.4"
+  }
+}
+----
+
+.stage configuration
+[source,hcl]
+----
+include "parent" {
+  path           = "/path/to/parent/configuration"
+  merge_strategy = "deep"
+}
+
+inputs = {
+  container_image = {
+    tag = "v0.0.3"
+  }
+}
+----
+
+In this way, you can leverage deep merge to refactor complex nested inputs in your Terragrunt configuration to further
+DRY up the config.
+
+
+[[expose_include]]
+===== Expose include
+
+Terragrunt offers the ability to reference values defined in parent configurations through exposed includes. This
+feature is enabled when the `expose = true` attribute is set on the `include` block. Exposed `include` blocks allow the
+child configuration to reference values that are defined in the parent configuration. These values are available with
+the reference `include.LABEL`.
+
+This is most useful to reference `locals` that are defined in the common component configuration to reuse them in the
+child.
+
+For example, if you wanted to test a new version of the module only in dev:
+
+.common component configuration
+[source,hcl]
+----
+locals {
+  source_base_url = "git::git@github.com:gruntwork-io/terraform-aws-service-catalog.git//modules/networking/vpc"
+}
+
+terraform {
+  source = "${local.source_base_url}?ref=v0.65.0"
+}
+----
+
+.dev configuration
+[source,hcl]
+----
+include "parent" {
+  path   = "/path/to/parent/configuration"
+  expose = true
+}
+
+terraform {
+  source = "${include.parent.locals.source_base_url}?ref=v0.66.0"
+}
+----
+
+Note that the availability of values is subject to the
+https://terragrunt.gruntwork.io/docs/etting-started/configuration/#configuration-parsing-order[configuration parsing
+order] of Terragrunt.
+This means that you won't be able to reference later stage values in early stage blocks, like accessing parent `inputs`
+in `locals`.
```

Comment:
Can you please add a code snippet under this paragraph that illustrates what you cannot do?

Reply:
Hmm I'm not sure about that. In general, I'm not a fan of putting in broken code snippets in a guide, as it's natural for people to focus on the code than the text. It's all too easy to scroll through, see the code snippet, and assume that works when the example is showing what you can't do.

Reply:
I added another textual example to hopefully clarify the point more: 08384da9b6b3d5d036121a22e0d84ac4f2aeb97f

Reply:
I can see your point. Let me comment on the latest code.

Reply:
I'll just add, seeing that example of what you can't do is _super_ helpful for me. I didn't understand fully what you meant until I saw that.

Reply:
I'm going to resolve this because I think being a little more verbose is enough to help here.